beps/0005-split-backend-discovery/README.md

```diff
@@ -186,13 +220,48 @@ paths:
             application/json:
               schema:
                 type: object
+                # using an object here so that we can add extra metadata like cache ttl, etc.
                 properties:
-                  internal:
+                  url:
                     type: string
-                  external:
+  /by-plugin/{pluginId}/external-base-url:
+    get:
+      summary: Get external registration URL for a given pluginId.
+      operationId: getExternalRegistrationByPlugin
+      parameters:
+        - name: pluginId
+          in: path
+          description: The plugin ID to get information for.
+          required: true
+          schema:
+            type: string
+      responses:
+        '200':
+          description: Success
+          content:
+            application/json:
+              schema:
+                type: object
```

Comment:
Hm. Aren't we doing essentially what DNS does, ie returning a randomized array of candidates? Seems unnecessary to ask this one over and over to get single URLs since the client side probably wants to apply a heuristic across a set of candidates (minimum-latency, round-robin, etc, plus how to do failovers).

Same obviously for the other method above

Reply:
Hmm, depends on how the internal routing is set up. My (maybe half-baked ?) assumption here is that this discovery service would sit on top existing routing set ups with load balancers. For example, catalog may have 4 containers, all sitting behind a load balancer with a single URL - that single URL would be registered and traffic would be routed there. I wasn't thinking this would be a complete routing solution in and of itself. 

Is full support for internal/external URLs something we'd want? ie just deploying a Backstage container with xyz credentials and URLs set up, will register it automatically with the gateway node and make it reachable from frontend and backend.

Reply:
Speaking of load balancing - how could the service that registers itself even know how to formulate the URLs to itself? Especially the external ones? It seems to me that registration could never be about more than IP addresses basically, and then the gateway and/or direct callers would have to know what to do in terms of reachability for those. Remember that it's open source plugins registering themselves. And the discovery service/api used to be the customization point where you put the knowledge of how to make URLs out of plugin IDs. Now you're handing off some of that to a gateway - There's some unclarity here as to what an adopter might do to adapt all of this to their own special circumstances.

Reply:
That same situation happens for non-load balanced ones too though, right? We don't currently have a great way to formulate remote internal URLs. The `HostDiscovery` just uses localhost or the statically defined discovery config.

Maybe it is time for a pivot to a true gateway design :) 

Gateway would be ideally colocated with the other backend services (though not necessarily). Registration would still want auth of some sort, I'd like that to be a second BEP though, since it seems fairly _involved_. Gateway nodes could host plugins but generally should just route to other backends. We'd want to have some interface for instances to get their own IP address/network address (EC2 metadata service/k8s layer), that would then be passed at registration time. 

Generally the routing would sit at the gateway node. Clients would just send requests to the gateway node(s), and the gateway node would relay the response. This kind of turns the gateway node into a load balancer and router, so it may make sense to support a routing abstraction here as well (round-robin, latency-based, etc).

Customization could happen at the edges, gateway -> backend, client -> gateway.

Reply:
Or... Well, it depends a bit on how the ip-to-url pluggability looks

Reply:
Yerp, I'll take another look at this tonight. The gateway layer seems more important than I had originally thought.